<a href="https://colab.research.google.com/github/VitorFRodrigues/Alura_Challenges/blob/main/notebook/Alura_Challenge_DataScience1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
dados_URL = 'https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json'

In [2]:
import pandas as pd
dados = pd.read_json(dados_URL)

In [3]:
dados

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

As últimas colunas estão compiladas com dicionários:

Vamos expandir as colunas com o auxilio da função *json_normalize*. Vamos criar uma função para automatizar

In [4]:
def normalizar(dados, colunas):
  for i in range(0, len(colunas)):
    normaliza = pd.json_normalize(dados[colunas[i]])
    dados = pd.concat([dados, normaliza], axis='columns').drop(columns=colunas[i])
  return dados

In [5]:
colunas = ['customer', 'phone', 'internet', 'account']
dados = normalizar(dados, colunas)

In [6]:
dados

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7267 rows x 21 columns]

#1.0 Analise e categorização dos dados

##1.1 CustomerID

Checagem se cada ID é único

In [7]:
len(dados['customerID'].unique())

7267

Valores únicos de customerID é igual ao número de linhas dos dados o que comprova que cada linha possui ID único.

##1.2 Churn

Checagem se opção de escolha se limita a Yes / No

In [8]:
dados['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

Vemos que algumas linhas não possuem tag, vamos identificar quantas linhas:

In [9]:
len(dados.query('Churn!="Yes" & Churn!="No"'))

224

In [10]:
len(dados.query('Churn!="Yes" & Churn!="No"'))/len(dados)*100

3.0824274115866244

São 224 linhas sem Tag o que corresponde a cerca de 3% de todo conjunto de dados. Estas serão armazenada em variável específica e removida do conjunto de dados

In [11]:
dados_sem_tag = dados.query('Churn!="Yes" & Churn!="No"')
dados_sem_tag

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30    0047-ZHDTW        Female              0      No         No      11   
75    0120-YZLQA          Male              0      No         No      71   
96    0154-QYHJU          Male              0      No         No      29   
98    0162-RZGMZ        Female              1      No         No       5   
175   0274-VVQOQ          Male              1     Yes         No      65   
...          ...   ...     ...            ...     ...        ...     ...   
7158  9840-GSRFX        Female              0      No         No      14   
7180  9872-RZQQB        Female              0     Yes         No      49   
7211  9920-GNDMB          Male              0      No         No       9   
7239  9955-RVWSC        Female              0     Yes        Yes      67   
7247  9966-VYRTZ        Female              0     Yes        Yes      31   

     PhoneService     MultipleLines InternetService  ...         OnlineBackup  \
30            Yes               Yes     Fiber optic  ...                   No   
75            Yes                No              No  ...  No internet service   
96            Yes                No             DSL  ...                  Yes   
98            Yes                No             DSL  ...                  Yes   
175           Yes               Yes     Fiber optic  ...                  Yes   
...           ...               ...             ...  ...                  ...   
7158          Yes               Yes             DSL  ...                  Yes   
7180           No  No phone service             DSL  ...                   No   
7211          Yes               Yes     Fiber optic  ...                   No   
7239          Yes                No              No  ...  No internet service   
7247          Yes                No              No  ...  No internet service   

         DeviceProtection          TechSupport          StreamingTV  \
30                     No                   No                   No   
75    No internet service  No internet service  No internet service   
96                     No                  Yes                   No   
98                     No                  Yes                   No   
175                   Yes                   No                  Yes   
...                   ...                  ...                  ...   
7158                   No                   No                   No   
7180                   No                   No                  Yes   
7211                   No                   No                   No   
7239  No internet service  No internet service  No internet service   
7247  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
30                     No  Month-to-month              Yes   
75    No internet service        Two year              Yes   
96                     No        One year              Yes   
98                     No  Month-to-month               No   
175                   Yes        One year              Yes   
...                   ...             ...              ...   
7158                   No        One year              Yes   
7180                   No  Month-to-month               No   
7211                   No  Month-to-month              Yes   
7239  No internet service        Two year              Yes   
7247  No internet service  Month-to-month              Yes   

                  PaymentMethod Charges.Monthly  Charges.Total  
30    Bank transfer (automatic)           79.00          929.3  
75      Credit card (automatic)           19.90         1355.1  
96             Electronic check           58.75         1696.2  
98      Credit card (automatic)           59.90         287.85  
175   Bank transfer (automatic)          103.15        6792.45  
...                         ...             ...            ...  
7158               Mailed check           54.25          773.2  
71

In [12]:
dados_limpos = dados.query('Churn=="Yes" or Churn=="No"')
dados_limpos

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0     0002-ORFBO    No  Female              0     Yes        Yes       9   
1     0003-MKNFE    No    Male              0      No         No       9   
2     0004-TLHLJ   Yes    Male              0      No         No       4   
3     0011-IGKFF   Yes    Male              1     Yes         No      13   
4     0013-EXCHZ   Yes  Female              1     Yes         No       3   
...          ...   ...     ...            ...     ...        ...     ...   
7262  9987-LUTYD    No  Female              0      No         No      13   
7263  9992-RRAMN   Yes    Male              0     Yes         No      22   
7264  9992-UJOEL    No    Male              0      No         No       2   
7265  9993-LHIEB    No    Male              0     Yes        Yes      67   
7266  9995-HOTOH    No    Male              0     Yes        Yes      63   

     PhoneService     MultipleLines InternetService  ... OnlineBackup  \
0             Yes                No             DSL  ...          Yes   
1             Yes               Yes             DSL  ...           No   
2             Yes                No     Fiber optic  ...           No   
3             Yes                No     Fiber optic  ...          Yes   
4             Yes                No     Fiber optic  ...           No   
...           ...               ...             ...  ...          ...   
7262          Yes                No             DSL  ...           No   
7263          Yes               Yes     Fiber optic  ...           No   
7264          Yes                No             DSL  ...          Yes   
7265          Yes                No             DSL  ...           No   
7266           No  No phone service             DSL  ...          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0                 Yes      Mailed check           65.60          593.3  
1                  No      Mailed check           59.90          542.4  
2                 Yes  Electronic check           73.90         280.85  
3                 Yes  Electronic check           98.00        1237.85  
4                 Yes      Mailed check           83.90          267.4  
...               ...               ...             ...            ...  
7262               No      Mailed check           55.15          742.9  
7263              Yes  Electronic check           85.10         1873.7  
7264              Yes      Mailed check           50.30          92.75  
7265               No      Mailed check           67.85        4627.65  
7266               No  Electronic check           59.00         3707.6  

[7043 rows x 21 columns]

In [13]:
dados_limpos['Churn'].unique()

array(['No', 'Yes'], dtype=object)

##1.3 Categorização de demais colunas

Muitas features possuem categorização simples (ex.: Yes/No, Female/Male, etc.). Sendo assim, vamos criar uma função que busca por tais colunas e categoriza-las.

In [14]:
for columns in dados_limpos:
  print(f'Coluna {columns}: {dados_limpos[columns].unique()}')

Coluna customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Coluna Churn: ['No' 'Yes']
Coluna gender: ['Female' 'Male']
Coluna SeniorCitizen: [0 1]
Coluna Partner: ['Yes' 'No']
Coluna Dependents: ['Yes' 'No']
Coluna tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42
  0]
Coluna PhoneService: ['Yes' 'No']
Coluna MultipleLines: ['No' 'Yes' 'No phone service']
Coluna InternetService: ['DSL' 'Fiber optic' 'No']
Coluna OnlineSecurity: ['No' 'Yes' 'No internet service']
Coluna OnlineBackup: ['Yes' 'No' 'No internet service']
Coluna DeviceProtection: ['No' 'Yes' 'No internet service']
Coluna TechSupport: ['Yes' 'No' 'No internet service']
Coluna StreamingTV: ['Yes' 'No' 'No internet service']
Coluna StreamingMovies: ['No' 'Yes' 'No internet service']
Coluna Contract: ['One yea

Camos categorizar itens simples utilizando função replace:

In [15]:
substituir = {'Female': 0,
              'Male': 1,
              'No': 0,
              'Yes': 1,
              'No phone service': 2,
              'No internet service': 3,
              'DSL': 4,
              'Fiber optic': 5,
              'One year': 6,
              'Month-to-month': 7,
              'Two year': 8,
              'Mailed check': 9,
              'Electronic check': 10,
              'Credit card (automatic)': 11,
              'Bank transfer (automatic)': 12,
              }
dados_categorizados = dados_limpos.replace(substituir)

In [16]:
for columns in dados_categorizados:
  print(f'Coluna {columns}: {dados_categorizados[columns].unique()}')

Coluna customerID: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']
Coluna Churn: [0 1]
Coluna gender: [0 1]
Coluna SeniorCitizen: [0 1]
Coluna Partner: [1 0]
Coluna Dependents: [1 0]
Coluna tenure: [ 9  4 13  3 71 63  7 65 54 72  5 56 34  1 45 50 23 55 26 69 37 49 66 67
 20 43 59 12 27  2 25 29 14 35 64 39 40 11  6 30 70 57 58 16 32 33 10 21
 61 15 44 22 24 19 47 62 46 52  8 60 48 28 41 53 68 31 36 17 18 51 38 42
  0]
Coluna PhoneService: [1 0]
Coluna MultipleLines: [0 1 2]
Coluna InternetService: [4 5 0]
Coluna OnlineSecurity: [0 1 3]
Coluna OnlineBackup: [1 0 3]
Coluna DeviceProtection: [0 1 3]
Coluna TechSupport: [1 0 3]
Coluna StreamingTV: [1 0 3]
Coluna StreamingMovies: [0 1 3]
Coluna Contract: [6 7 8]
Coluna PaperlessBilling: [1 0]
Coluna PaymentMethod: [ 9 10 11 12]
Coluna Charges.Monthly: [65.6  59.9  73.9  ... 91.75 68.8  67.85]
Coluna Charges.Total: ['593.3' '542.4' '280.85' ... '742.9' '4627.65' '3707.6']


Vamos buscar quais dados se encontram não-categorizados ou não-numéricos. A expectativa é que apenas o 'customerID' esteja não-categorizada:

In [17]:
dados_categorizados.describe().columns
for columns in dados_categorizados:
  if columns not in dados_categorizados.describe().columns:
    print(columns)

customerID
Charges.Total


Vamos transformar Charges.Total em número:

In [18]:
dados_categorizados['Charges.Total'] = pd.to_numeric(dados_categorizados['Charges.Total'], errors='coerce')

Agora podemos adquirir informações do data frame com a função describe

In [19]:
dados_categorizados.describe()

Churn       gender  SeniorCitizen      Partner   Dependents  \
count  7043.000000  7043.000000    7043.000000  7043.000000  7043.000000   
mean      0.265370     0.504756       0.162147     0.483033     0.299588   
std       0.441561     0.500013       0.368612     0.499748     0.458110   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000     1.000000       0.000000     0.000000     0.000000   
75%       1.000000     1.000000       0.000000     1.000000     1.000000   
max       1.000000     1.000000       1.000000     1.000000     1.000000   

            tenure  PhoneService  MultipleLines  InternetService  \
count  7043.000000   7043.000000    7043.000000      7043.000000   
mean     32.371149      0.903166       0.615505         3.572909   
std      24.559481      0.295752       0.656039         1.929874   
min       0.000000      0.000000       0.000000         0.000000   
25%       9.000000      1.000000       0.000000         4.000000   
50%      29.000000      1.000000       1.000000         4.000000   
75%      55.000000      1.000000       1.000000         5.000000   
max      72.000000      1.000000       2.000000         5.000000   

       OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
count     7043.000000   7043.000000       7043.000000  7043.000000   
mean         0.936675      0.994889          0.993895     0.940224   
std          1.165986      1.142491          1.142921     1.164650   
min          0.000000      0.000000          0.000000     0.000000   
25%          0.000000      0.000000          0.000000     0.000000   
50%          1.000000      1.000000          1.000000     1.000000   
75%          1.000000      1.000000          1.000000     1.000000   
max          3.000000      3.000000          3.000000     3.000000   

       StreamingTV  StreamingMovies     Contract  PaperlessBilling  \
count  7043.000000      7043.000000  7043.000000       7043.000000   
mean      1.034360         1.037910     7.031521          0.592219   
std       1.124568         1.122874     0.669984          0.491457   
min       0.000000         0.000000     6.000000          0.000000   
25%       0.000000         0.000000     7.000000          0.000000   
50%       1.000000         1.000000     7.000000          1.000000   
75%       1.000000         1.000000     7.000000          1.000000   
max       3.000000         3.000000     8.000000          1.000000   

       PaymentMethod  Charges.Monthly  Charges.Total  
count    7043.000000      7043.000000    7032.000000  
mean       10.425671        64.761692    2283.300441  
std         1.068104        30.090047    2266.771362  
min         9.000000        18.250000      18.800000  
25%        10.000000        35.500000     401.450000  
50%        10.000000        70.350000    1397.475000  
75%        11.000000        89.850000    3794.737500  
max        12.000000       118.750000    8684.800000

Vemos que coluna Charges.Total possui contagem diferente das demais colunas. Por regra, cada coluna de dataframe possui tipo definido. Sendo assim, há chances destes números serem do tipo NaN (Not-a-number).

In [20]:
dados_categorizados[dados_categorizados.isna().any(axis=1)]

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
975   1371-DWPAZ      0       0              0        1           1       0   
1775  2520-SGTTA      0       0              0        1           1       0   
1955  2775-SEFEE      0       1              0        0           1       0   
2075  2923-ARZLG      0       1              0        1           1       0   
2232  3115-CZMZD      0       1              0        0           1       0   
2308  3213-VVOLG      0       1              0        1           1       0   
2930  4075-WKNIU      0       0              0        1           1       0   
3134  4367-NUYAO      0       1              0        1           1       0   
3203  4472-LVYGI      0       0              0        1           1       0   
4169  5709-LVOEQ      0       0              0        1           1       0   
5599  7644-OMVMY      0       1              0        1           1       0   

      PhoneService  MultipleLines  InternetService  ...  OnlineBackup  \
975              0              2                4  ...             1   
1775             1              0                0  ...             3   
1955             1              1                4  ...             1   
2075             1              0                0  ...             3   
2232             1              0                0  ...             3   
2308             1              1                0  ...             3   
2930             1              1                4  ...             1   
3134             1              1                0  ...             3   
3203             0              2                4  ...             0   
4169             1              0                4  ...             1   
5599             1              0                0  ...             3   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
975                  1            1            1                0         8   
1775                 3            3            3                3         8   
1955                 0            1            0                0         8   
2075                 3            3            3                3         6   
2232                 3            3            3                3         8   
2308                 3            3            3                3         8   
2930                 1            1            1                0         8   
3134                 3            3            3                3         8   
3203                 1            1            1                0         8   
4169                 1            0            1                1         8   
5599                 3            3            3                3         8   

      PaperlessBilling  PaymentMethod  Charges.Monthly  Charges.Total  
975                  0             11            56.05            NaN  
1775                 0              9            20.00            NaN  
1955                 1             12            61.90            NaN  
2075                 1              9            19.70            NaN  
2232                 0              9            20.25            NaN  
2308                 0              9            25.35            NaN  
2930                 0              9            73.35            NaN  
3134                 0              9            25.75            NaN  
3203                 1             12            52.55            NaN  
4169                 0              9            80.85            NaN  
5599                 0              9            19.85            NaN  

[11 rows x 21 columns]

Como 11 linhas representa apenas 0,15% dentro das 7.043 linhas, vamos remove-las do dataframe 

In [21]:
dados_NaN = dados_categorizados[dados_categorizados.isna().any(axis=1)]
dados_categorizados = dados_categorizados.drop(dados_categorizados[dados_categorizados.isna().any(axis=1)].index)

In [22]:
dados_categorizados.describe()

Churn       gender  SeniorCitizen      Partner   Dependents  \
count  7032.000000  7032.000000    7032.000000  7032.000000  7032.000000   
mean      0.265785     0.504693       0.162400     0.482509     0.298493   
std       0.441782     0.500014       0.368844     0.499729     0.457629   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000     1.000000       0.000000     0.000000     0.000000   
75%       1.000000     1.000000       0.000000     1.000000     1.000000   
max       1.000000     1.000000       1.000000     1.000000     1.000000   

            tenure  PhoneService  MultipleLines  InternetService  \
count  7032.000000   7032.000000    7032.000000      7032.000000   
mean     32.421786      0.903299       0.615330         3.575654   
std      24.545260      0.295571       0.655867         1.928525   
min       1.000000      0.000000       0.000000         0.000000   
25%       9.000000      1.000000       0.000000         4.000000   
50%      29.000000      1.000000       1.000000         4.000000   
75%      55.000000      1.000000       1.000000         5.000000   
max      72.000000      1.000000       2.000000         5.000000   

       OnlineSecurity  OnlineBackup  DeviceProtection  TechSupport  \
count     7032.000000   7032.000000       7032.000000  7032.000000   
mean         0.935011      0.993316          0.992321     0.938567   
std          1.165284      1.141820          1.142250     1.163950   
min          0.000000      0.000000          0.000000     0.000000   
25%          0.000000      0.000000          0.000000     0.000000   
50%          1.000000      1.000000          1.000000     1.000000   
75%          1.000000      1.000000          1.000000     1.000000   
max          3.000000      3.000000          3.000000     3.000000   

       StreamingTV  StreamingMovies     Contract  PaperlessBilling  \
count  7032.000000      7032.000000  7032.000000       7032.000000   
mean      1.032850         1.036832     7.030290          0.592719   
std       1.123912         1.122016     0.669398          0.491363   
min       0.000000         0.000000     6.000000          0.000000   
25%       0.000000         0.000000     7.000000          0.000000   
50%       1.000000         1.000000     7.000000          1.000000   
75%       1.000000         1.000000     7.000000          1.000000   
max       3.000000         3.000000     8.000000          1.000000   

       PaymentMethod  Charges.Monthly  Charges.Total  
count    7032.000000      7032.000000    7032.000000  
mean       10.426763        64.798208    2283.300441  
std         1.067504        30.085974    2266.771362  
min         9.000000        18.250000      18.800000  
25%        10.000000        35.587500     401.450000  
50%        10.000000        70.350000    1397.475000  
75%        11.000000        89.862500    3794.737500  
max        12.000000       118.750000    8684.800000

Agora que os dados estão devidamente categorizados, vamos descrever as colunas para entendimento do pessoal da produção:

In [23]:
descricao_colunas = {
    'customerID': 'número de identificação único de cada cliente',
    'Churn': 'se o cliente deixou ou não a empresa',
    'gender': 'gênero (masculino e feminino)',
    'SeniorCitizen': 'informação sobre um cliente ter ou não idade igual ou maior que 65 anos',
    'Partner': 'se o cliente possui ou não um parceiro ou parceira',
    'Dependents': 'se o cliente possui ou não dependentes',
    'tenure': 'meses de contrato do cliente',
    'PhoneService': 'assinatura de serviço telefônico',
    'MultipleLines': 'assisnatura de mais de uma linha de telefone',
    'InternetService': 'assinatura de um provedor internet',
    'OnlineSecurity': 'assinatura adicional de segurança online',
    'OnlineBackup': 'assinatura adicional de backup online',
    'DeviceProtection': 'assinatura adicional de proteção no dispositivo',
    'TechSupport': 'assinatura adicional de suporte técnico, menos tempo de espera',
    'StreamingTV': 'assinatura de TV a cabo',
    'StreamingMovies': 'assinatura de streaming de filmes',
    'Contract': 'tipo de contrato',
    'PaperlessBilling': 'se o cliente prefere receber online a fatura',
    'PaymentMethod': 'forma de pagamento',
    'Charges.Monthly': 'total de todos os serviços do cliente por mês',
    'Charges.Total': 'total gasto pelo cliente'}

Para finalizar o tratamento, vamos adicionar uma coluna indicando o custo diário de cada cliente

In [27]:
#Considerando a media mensal de 30 dias, temos
dados_categorizados['Charges.Daily'] = dados_categorizados['Charges.Monthly']/30
dados_categorizados.head()

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  0002-ORFBO      0       0              0        1           1       9   
1  0003-MKNFE      0       1              0        0           0       9   
2  0004-TLHLJ      1       1              0        0           0       4   
3  0011-IGKFF      1       1              1        1           0      13   
4  0013-EXCHZ      1       0              1        1           0       3   

   PhoneService  MultipleLines  InternetService  ...  DeviceProtection  \
0             1              0                4  ...                 0   
1             1              1                4  ...                 0   
2             1              0                5  ...                 1   
3             1              0                5  ...                 1   
4             1              0                5  ...                 0   

   TechSupport  StreamingTV  StreamingMovies  Contract  PaperlessBilling  \
0            1            1                0         6                 1   
1            0            0                1         7                 0   
2            0            0                0         7                 1   
3            0            1                1         7                 1   
4            1            1                0         7                 1   

   PaymentMethod  Charges.Monthly  Charges.Total  Charges.Daily  
0              9             65.6         593.30       2.186667  
1              9             59.9         542.40       1.996667  
2             10             73.9         280.85       2.463333  
3             10             98.0        1237.85       3.266667  
4              9             83.9         267.40       2.796667  

[5 rows x 22 columns]

Vamos reorganizar a ordem das colunas

In [28]:
dados_categorizados.columns.tolist()

['customerID',
 'Churn',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'Charges.Monthly',
 'Charges.Total',
 'Charges.Daily']

In [30]:
rearranjo = ['customerID', 'Churn', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 
             'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'Charges.Daily', 'Charges.Monthly', 'Charges.Total']
dados_categorizados = dados_categorizados[rearranjo]
dados_categorizados.head()

customerID  Churn  gender  SeniorCitizen  Partner  Dependents  tenure  \
0  0002-ORFBO      0       0              0        1           1       9   
1  0003-MKNFE      0       1              0        0           0       9   
2  0004-TLHLJ      1       1              0        0           0       4   
3  0011-IGKFF      1       1              1        1           0      13   
4  0013-EXCHZ      1       0              1        1           0       3   

   PhoneService  MultipleLines  InternetService  ...  DeviceProtection  \
0             1              0                4  ...                 0   
1             1              1                4  ...                 0   
2             1              0                5  ...                 1   
3             1              0                5  ...                 1   
4             1              0                5  ...                 0   

   TechSupport  StreamingTV  StreamingMovies  Contract  PaperlessBilling  \
0            1            1                0         6                 1   
1            0            0                1         7                 0   
2            0            0                0         7                 1   
3            0            1                1         7                 1   
4            1            1                0         7                 1   

   PaymentMethod  Charges.Daily  Charges.Monthly  Charges.Total  
0              9       2.186667             65.6         593.30  
1              9       1.996667             59.9         542.40  
2             10       2.463333             73.9         280.85  
3             10       3.266667             98.0        1237.85  
4              9       2.796667             83.9         267.40  

[5 rows x 22 columns]

#2.0 Analise exploratória e data vizualization